pkg/controller/awsmachine/awsmachine_controller.go

```diff
@@ -153,116 +153,71 @@ func (r *ReconcileAWSMachine) Reconcile(request reconcile.Request) (reconcile.Re
-		scope.SetProviderID(fmt.Sprintf("aws:////%s", *scope.GetInstanceID()))
-	}
-
-	if exist {
-		scope.Info("Reconciling Machine triggers idempotent update")
-		return r.update(scope)
-	}
-
-	// Machine resource created. Machine does not yet exist.
-	scope.Info("Reconciling Machine triggers idempotent create")
-	return r.create(scope)
-}
-
-// create creates a machine and is invoked by the machine controller.
-func (r *ReconcileAWSMachine) create(scope *scope.MachineScope) error {
 	if scope.Parent.Cluster.Annotations[infrav1.AnnotationClusterInfrastructureReady] != infrav1.ValueReady {
 		scope.Info("Cluster infrastructure is not ready yet - requeuing machine")
 		return &capierrors.RequeueAfterError{RequeueAfter: waitForClusterInfrastructureReadyDuration}
 	}
 
 	ec2svc := ec2.NewService(scope.Parent)
 
-	scope.Info("Retrieving machines for cluster")
-	machineList := &clusterv1.MachineList{}
-	if err := r.List(context.Background(), machineList, scope.Parent.ListOptionsLabelSelector()); err != nil {
-		return errors.Wrapf(err, "failed to retrieve machines in cluster %q", scope.Parent.Name())
-	}
-
-	controlPlaneMachines := util.GetControlPlaneMachinesFromList(machineList)
-	if len(controlPlaneMachines) == 0 {
-		scope.Info("No control plane machines exist yet - requeuing")
-		return &capierrors.RequeueAfterError{RequeueAfter: waitForControlPlaneMachineExistenceDuration}
-	}
-
-	// Create the Machine.
-	i, err := ec2svc.CreateOrGetMachine(scope)
+	// Look for the instance by tags first.
+	instance, err := ec2svc.InstanceByTags(scope)
 	if err != nil {
-		return errors.Errorf("failed to create or get machine: %+v", err)
-	}
-
-	scope.SetInstanceID(i.ID)
-	scope.SetInstanceState(i.State)
-	scope.SetAnnotation("cluster-api-provider-aws", "true")
-
-	if err := r.reconcileLBAttachment(scope, i); err != nil {
-		return errors.Errorf("failed to reconcile LB attachment: %+v", err)
+		return errors.Wrapf(err, "failed to create AWSMachine instance")
 	}
 
-	scope.Info("Create completed")
-	return nil
-}
-
-func (r *ReconcileAWSMachine) exists(scope *scope.MachineScope) (bool, error) {
-	ec2svc := ec2.NewService(scope.Parent)
-
-	// TODO worry about pointers. instance if exists returns *any* instance
-	if scope.GetInstanceID() == nil {
-		return false, nil
-	}
-
-	instance, err := ec2svc.InstanceIfExists(scope.GetInstanceID())
-	if err != nil {
-		return false, errors.Errorf("failed to retrieve instance: %+v", err)
-	}
+	if pid := scope.GetProviderID(); pid == nil || *pid == "" {
+		// If the Spec.ProviderID is not set and the instance is not found,
+		// create the AWSMachine.
+		if instance == nil {
+			instance, err = ec2svc.CreateInstance(scope)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create AWSMachine instance")
+			}
+		}
 
-	if instance == nil {
-		return false, nil
+		// Set the ProviderID.
+		scope.SetProviderID(fmt.Sprintf("aws:////%s", instance.ID))
+	} else if instance == nil {
```

Comment:
I'd like to try to find a way to make the branching and conditionals easier to follow. I don't have a good idea in my head yet. I'll keep thinking about it.

Reply:
I could move it to its own function, but then we'd get into the `getOrCreate` scenario again

Reply:
I think a `getOrCreate` scenario in this file is probably fine, but I'd like to avoid having something buried in the service code to make generating events cleaner.